.github/workflows/infra_plan.yaml

```
@@ -112,16 +112,26 @@ jobs:
         working-directory: ${{ steps.directory.outputs.dir }}
         run: |
           terraform plan -lock-timeout=3000s -no-color 2>&1 | grep -v "hidden-link:"  | tee plan_output.txt
+          EXIT_CODE=${PIPESTATUS[0]}
 
           OUTPUT=$(grep -Ev "Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
 
           printf "%s" "$OUTPUT" > plan_output_multiline.txt
 
-          if grep -q "::error::Terraform exited with code" plan_output.txt; then
-            echo "failed"
+          # exit 1 if the plan failed
+          if [ "$EXIT_CODE" -ne 0 ]; then
+            echo "Terraform plan failed with exit code $EXIT_CODE"
             exit 1
           fi
 
+      # Upload full Terraform plan output as artifact
+      - name: Upload Terraform Plan Artifact
+        if: always() && github.event_name == 'pull_request'
+        uses: actions/upload-artifact@v3
```

If you agree, I would add an id to the step so we can get its output in the following step
```suggestion
        if: always() && github.event_name == 'pull_request'
        id: artifact-upload-step
        uses: actions/upload-artifact@v3
```